File: Pokemon SwiftUI/Pokemon SwiftUI/Pokemon/Pokemon.swift
//
//  Pokemon.swift
//  Pokemon SwiftUI
//
//  Created by Joshua Homann on 6/20/19.
//  Copyright © 2019 com.josh. All rights reserved.
//

import Foundation
import SwiftUI

// MARK: - Pokemon
struct Pokemon: Codable {
  let id: String
  let pkdxID, nationalID: Int
  let name: String
  let v: Int
  let imageURL: URL
  let pokemonDescription: String
  let artURL: URL
  let types: [String]
  let evolutions: [Evolution]

  enum CodingKeys: String, CodingKey {
    case id = "_id"
    case pkdxID = "pkdx_id"
    case nationalID = "national_id"
    case name
    case v = "__v"
    case imageURL = "image_url"
    case pokemonDescription = "description"
    case artURL = "art_url"
    case types, evolutions
  }
  static let all: [Pokemon] = (Bundle.main.url(forResource: "pokemon", withExtension: "json")
    .flatMap {try! Data(contentsOf: $0)}
    .flatMap {try! JSONDecoder().decode([Pokemon].self, from: $0)} ?? [])
    .sorted { $0.name < $1.name }

}

// MARK: - Evolution
struct Evolution: Codable {
  let level: Int?
  let method: Method
  let to, id: String

  enum CodingKeys: String, CodingKey {
    case level, method, to
    case id = "_id"
  }
}

enum Method: String, Codable {
  case levelUp = "level_up"
  case other = "other"
  case stone = "stone"
  case trade = "trade"
}

File: pokemon_flutter/lib/pokemon.dart
import 'dart:convert';
import 'dart:async' show Future;
import 'package:flutter/services.dart' show rootBundle;


List<Pokemon> pokemonFromJson(String str) => new List<Pokemon>.from(json.decode(str).map((x) => Pokemon.fromJson(x)));

String pokemonToJson(List<Pokemon> data) => json.encode(new List<dynamic>.from(data.map((x) => x.toJson())));

Future<List<Pokemon>> loadAllPokemon() async {
  final contents = await rootBundle.loadString('assets/pokemon.json');
  return pokemonFromJson(contents);
}

class Pokemon {

  String id;
  int pkdxId;
  int nationalId;
  String name;
  int v;
  String imageUrl;
  String description;
  String artUrl;
  List<String> types;
  List<Evolution> evolutions;

  Pokemon({
    this.id,
    this.pkdxId,
    this.nationalId,
    this.name,
    this.v,
    this.imageUrl,
    this.description,
    this.artUrl,
    this.types,
    this.evolutions,
  });

  factory Pokemon.fromJson(Map<String, dynamic> json) => new Pokemon(
    id: json["_id"],
    pkdxId: json["pkdx_id"],
    nationalId: json["national_id"],
    name: json["name"],
    v: json["__v"],
    imageUrl: json["image_url"],
    description: json["description"],
    artUrl: json["art_url"],
    types: new List<String>.from(json["types"].map((x) => x)),
    evolutions: new List<Evolution>.from(json["evolutions"].map((x) => Evolution.fromJson(x))),
  );

  Map<String, dynamic> toJson() => {
    "_id": id,
    "pkdx_id": pkdxId,
    "national_id": nationalId,
    "name": name,
    "__v": v,
    "image_url": imageUrl,
    "description": description,
    "art_url": artUrl,
    "types": new List<dynamic>.from(types.map((x) => x)),
    "evolutions": new List<dynamic>.from(evolutions.map((x) => x.toJson())),
  };
}

class Evolution {
  int level;
  Method method;
  String to;
  String id;

  Evolution({
    this.level,
    this.method,
    this.to,
    this.id,
  });

  factory Evolution.fromJson(Map<String, dynamic> json) => new Evolution(
    level: json["level"] == null ? null : json["level"],
    method: methodValues.map[json["method"]],
    to: json["to"],
    id: json["_id"],
  );

  Map<String, dynamic> toJson() => {
    "level": level == null ? null : level,
    "method": methodValues.reverse[method],
    "to": to,
    "_id": id,
  };
}

enum Method { LEVEL_UP, OTHER, STONE, TRADE }

final methodValues = new EnumValues({
  "level_up": Method.LEVEL_UP,
  "other": Method.OTHER,
  "stone": Method.STONE,
  "trade": Method.TRADE
});

class EnumValues<T> {
  Map<String, T> map;
  Map<T, String> reverseMap;

  EnumValues(this.map);

  Map<T, String> get reverse {
    if (reverseMap == null) {
      reverseMap = map.map((k, v) => new MapEntry(v, k));
    }
    return reverseMap;
  }
}